File: cosmos_on_near/internal/staking/staking.go
package staking

import (
	"cosmos_on_near/internal/bank"
	"cosmos_on_near/internal/near"
	"cosmos_on_near/internal/storage"
	"strconv"
)

const (
	UnbondingPeriod    = 100 // blocks
	RewardPercentage   = 5   // 5% per reward cycle
)

type StakingModule struct {
	store      *storage.ModuleStore
	bankModule *bank.BankModule
}

func NewStakingModule(store storage.Store, bankModule *bank.BankModule) *StakingModule {
	return &StakingModule{
		store:      storage.NewModuleStore(store, "staking"),
		bankModule: bankModule,
	}
}

func (sm *StakingModule) getValidator(address string) *Validator {
	key := []byte("validator:" + address)
	data := sm.store.Get(key)
	return DeserializeValidator(data)
}

func (sm *StakingModule) setValidator(validator *Validator) {
	key := []byte("validator:" + validator.Address)
	sm.store.Set(key, validator.Serialize())
}

func (sm *StakingModule) getDelegation(delegator, validator string) *Delegation {
	key := []byte("delegation:" + delegator + ":" + validator)
	data := sm.store.Get(key)
	return DeserializeDelegation(data)
}

func (sm *StakingModule) setDelegation(delegation *Delegation) {
	key := []byte("delegation:" + delegation.Delegator + ":" + delegation.Validator)
	sm.store.Set(key, delegation.Serialize())
}

func (sm *StakingModule) addUnbondingEntry(entry *UnbondingEntry) {
	key := []byte("unbonding:" + strconv.FormatUint(entry.UnlockHeight, 10) + ":" + entry.Delegator + ":" + entry.Validator)
	sm.store.Set(key, entry.Serialize())
}

func (sm *StakingModule) Delegate(validatorAddr string, amount uint64) {
	delegator := "system" // TODO: Get from NEAR context
	validator := sm.getValidator(validatorAddr)
	if validator == nil {
		panic("Validator not found")
	}
	
	if !validator.IsActive {
		panic("Validator is not active")
	}
	
	sm.bankModule.Transfer(delegator, "staking_pool.testnet", amount)
	
	delegation := sm.getDelegation(delegator, validatorAddr)
	if delegation == nil {
		delegation = &Delegation{
			Delegator: delegator,
			Validator: validatorAddr,
			Amount:    0,
		}
	}
	
	delegation.Amount += amount
	validator.DelegatedStake += amount
	
	sm.setDelegation(delegation)
	sm.setValidator(validator)
	
	logMsg := "Delegated: " + delegator + " -> " + validatorAddr + " amount: " + strconv.FormatUint(amount, 10)
	near.LogString(logMsg)
}

func (sm *StakingModule) Undelegate(validatorAddr string, amount uint64) {
	delegator := "system" // TODO: Get from NEAR context
	currentHeight := sm.store.GetBlockHeight()
	
	delegation := sm.getDelegation(delegator, validatorAddr)
	if delegation == nil || delegation.Amount < amount {
		return // TODO: Handle error properly
	}
	
	validator := sm.getValidator(validatorAddr)
	if validator == nil {
		return // TODO: Handle error properly
	}
	
	delegation.Amount -= amount
	validator.DelegatedStake -= amount
	
	sm.setDelegation(delegation)
	sm.setValidator(validator)
	
	unbondingEntry := &UnbondingEntry{
		Delegator:    delegator,
		Validator:    validatorAddr,
		Amount:       amount,
		UnlockHeight: currentHeight + UnbondingPeriod,
	}
	
	sm.addUnbondingEntry(unbondingEntry)
	
	near.LogString("Undelegated: " + delegator + " from " + validatorAddr + " amount: " + strconv.FormatUint(amount, 10))
}

func (sm *StakingModule) AddValidator(address string) {
	validator := &Validator{
		Address:        address,
		DelegatedStake: 0,
		IsActive:       true,
	}
	
	sm.setValidator(validator)
	near.LogString("Added validator: " + address)
}

func (sm *StakingModule) BeginBlock(height uint64) {
	near.LogString("BeginBlock: " + strconv.FormatUint(height, 10))
}

func (sm *StakingModule) EndBlock(height uint64) {
	sm.processUnbonding(height)
	sm.distributeRewards()
	near.LogString("EndBlock: " + strconv.FormatUint(height, 10))
}

func (sm *StakingModule) processUnbonding(currentHeight uint64) {
	prefix := []byte("unbonding:")
	
	sm.store.IterPrefix(prefix, func(key, value []byte) bool {
		entry := DeserializeUnbondingEntry(value)
		if entry != nil && entry.UnlockHeight <= currentHeight {
			sm.bankModule.Transfer("staking_pool.testnet", entry.Delegator, entry.Amount)
			sm.store.Delete(key)
			near.LogString("Released unbonding: " + entry.Delegator + " amount: " + strconv.FormatUint(entry.Amount, 10))
		}
		return true
	})
}

func (sm *StakingModule) distributeRewards() {
	prefix := []byte("validator:")
	
	sm.store.IterPrefix(prefix, func(key, value []byte) bool {
		validator := DeserializeValidator(value)
		if validator != nil && validator.IsActive && validator.DelegatedStake > 0 {
			rewardAmount := validator.DelegatedStake * RewardPercentage / 100
			sm.bankModule.Mint("staking_pool.testnet", rewardAmount)
			near.LogString("Distributed rewards to validator: " + validator.Address + " amount: " + strconv.FormatUint(rewardAmount, 10))
		}
		return true
	})
}

File: cosmos_on_near_rust/src/staking.rs
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
use near_sdk::collections::UnorderedMap;
use near_sdk::{env, AccountId};
use crate::Balance;
use crate::bank::BankModule;

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Validator {
    pub address: AccountId,
    pub is_active: bool,
    pub total_delegated: Balance,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Delegation {
    pub validator: AccountId,
    pub amount: Balance,
}

#[derive(BorshDeserialize, BorshSerialize, Clone)]
pub struct UnbondingEntry {
    pub validator: AccountId,
    pub amount: Balance,
    pub release_height: u64,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct StakingModule {
    validators: UnorderedMap<AccountId, Validator>,
    delegations: UnorderedMap<AccountId, Vec<Delegation>>,
    unbonding_queue: UnorderedMap<AccountId, Vec<UnbondingEntry>>,
}

impl StakingModule {
    pub fn new() -> Self {
        Self {
            validators: UnorderedMap::new(b"v".to_vec()),
            delegations: UnorderedMap::new(b"d".to_vec()),
            unbonding_queue: UnorderedMap::new(b"u".to_vec()),
        }
    }

    pub fn add_validator(&mut self, validator_address: &AccountId) {
        let validator = Validator {
            address: validator_address.clone(),
            is_active: true,
            total_delegated: 0,
        };
        
        self.validators.insert(validator_address, &validator);
        env::log_str(&format!("Staking: Added validator {}", validator_address));
    }

    pub fn delegate(&mut self, delegator: &AccountId, validator: &AccountId, amount: Balance, bank: &mut BankModule) {
        // Check if validator exists
        let mut validator_info = self.validators.get(validator)
            .expect("Validator not found");
        
        // Check if delegator has sufficient balance
        assert!(bank.has_balance(delegator, amount), "Insufficient balance for delegation");
        
        // Burn tokens from delegator (move to staking pool)
        bank.burn(delegator, amount);
        
        // Update validator total
        validator_info.total_delegated += amount;
        self.validators.insert(validator, &validator_info);
        
        // Update delegator's delegations
        let mut delegations = self.delegations.get(delegator).unwrap_or_default();
        
        // Find existing delegation or create new one
        if let Some(existing) = delegations.iter_mut().find(|d| d.validator == *validator) {
            existing.amount += amount;
        } else {
            delegations.push(Delegation {
                validator: validator.clone(),
                amount,
            });
        }
        
        self.delegations.insert(delegator, &delegations);
        env::log_str(&format!("Staking: Delegated {} to {} from {}", amount, validator, delegator));
    }

    pub fn undelegate(&mut self, delegator: &AccountId, validator: &AccountId, amount: Balance, current_height: u64) {
        let mut delegations = self.delegations.get(delegator)
            .expect("No delegations found");
        
        // Find and update delegation
        let delegation = delegations.iter_mut()
            .find(|d| d.validator == *validator)
            .expect("Delegation not found");
        
        assert!(delegation.amount >= amount, "Insufficient delegated amount");
        
        // Update delegation
        delegation.amount -= amount;
        if delegation.amount == 0 {
            delegations.retain(|d| d.validator != *validator);
        }
        
        if delegations.is_empty() {
            self.delegations.remove(delegator);
        } else {
            self.delegations.insert(delegator, &delegations);
        }
        
        // Update validator
        let mut validator_info = self.validators.get(validator).unwrap();
        validator_info.total_delegated -= amount;
        self.validators.insert(validator, &validator_info);
        
        // Add to unbonding queue (100 blocks unbonding period)
        let mut unbonding = self.unbonding_queue.get(delegator).unwrap_or_default();
        unbonding.push(UnbondingEntry {
            validator: validator.clone(),
            amount,
            release_height: current_height + 100,
        });
        self.unbonding_queue.insert(delegator, &unbonding);
        
        env::log_str(&format!("Staking: Undelegated {} from {} by {}", amount, validator, delegator));
    }

    pub fn begin_block(&mut self, _height: u64) {
        // Process any begin block logic for staking
    }

    pub fn end_block(&mut self, height: u64, bank: &mut BankModule) {
        // Process unbonding queue
        self.process_unbonding_queue(height, bank);
        
        // Distribute rewards (5% flat rate)
        self.distribute_rewards(bank);
    }

    fn process_unbonding_queue(&mut self, current_height: u64, bank: &mut BankModule) {
        let mut accounts_to_update = Vec::new();
        
        for (account, unbonding_entries) in self.unbonding_queue.iter() {
            let mut updated_entries = Vec::new();
            let mut released_amount = 0;
            
            for entry in &unbonding_entries {
                if entry.release_height <= current_height {
                    released_amount += entry.amount;
                    env::log_str(&format!("Staking: Released {} to {} at height {}", 
                        entry.amount, account, current_height));
                } else {
                    updated_entries.push(entry.clone());
                }
            }
            
            if released_amount > 0 {
                bank.mint(&account, released_amount);
            }
            
            accounts_to_update.push((account.clone(), updated_entries));
        }
        
        // Update unbonding queue
        for (account, updated_entries) in accounts_to_update {
            if updated_entries.is_empty() {
                self.unbonding_queue.remove(&account);
            } else {
                self.unbonding_queue.insert(&account, &updated_entries);
            }
        }
    }

    fn distribute_rewards(&mut self, bank: &mut BankModule) {
        // 5% reward rate per block for all delegators
        let reward_rate = 5; // 5%
        
        for (delegator, delegations) in self.delegations.iter() {
            for delegation in &delegations {
                let reward = delegation.amount * reward_rate / 100;
                if reward > 0 {
                    bank.mint(&delegator, reward);
                    env::log_str(&format!("Staking: Rewarded {} to {} for delegation to {}", 
                        reward, delegator, delegation.validator));
                }
            }
        }
    }
}